onnxruntime/core/providers/cuda/reduction/reduction_ops.cc

```diff
@@ -601,32 +593,46 @@ Status ReduceKernel<allow_multi_axes>::ComputeImpl(OpKernelContext* ctx, cudnnRe
   return Status::OK();
 }
 
+template <bool allow_multi_axes>
+template <typename T, cudnnReduceTensorIndices_t ReduceTensorIndices>
+Status ReduceKernel<allow_multi_axes>::ComputeImpl(OpKernelContext* ctx, cudnnReduceTensorOp_t cudnn_reduce_op) const {
+  const Tensor* X = ctx->Input<Tensor>(0);
+
+  PrepareReduceMetadata prepare_reduce_metadata;
+  ORT_RETURN_IF_ERROR(PrepareForReduce(X,
+                                       keepdims_,
+                                       axes_,
+                                       prepare_reduce_metadata));
+
+  Tensor* Y = ctx->Output(0, prepare_reduce_metadata.squeezed_output_dims);
+
+  return ReduceComputeCore<T, ReduceTensorIndices>(*cuda_ep_, *X, prepare_reduce_metadata, *Y, cudnn_reduce_op, axes_,
+                                                   calculate_log_, calculate_sqt_, log_sum_exp_, fast_reduction_);
+}
+
 template <>
 template <>
 Status ReduceKernel<true>::ComputeImpl<int32_t, CUDNN_REDUCE_TENSOR_NO_INDICES>(OpKernelContext* ctx, cudnnReduceTensorOp_t cudnn_reduce_op) const {
   typedef typename ToCudaType<int32_t>::MappedType CudaT;
 
-  const Tensor* X = nullptr;
-  Tensor* Y = nullptr;
-
-  int64_t input_count = 0;
-  int64_t output_count = 0;
-  std::vector<int64_t> output_dims;
-  std::vector<int64_t> input_dims_cudnn;
-  std::vector<int64_t> output_dims_cudnn;
-  int64_t rank = 0;
-  int64_t stride = 0;
-  ORT_RETURN_IF_ERROR(PrepareForReduce(ctx,
+  const Tensor* X = ctx->Input<Tensor>(0);
+  PrepareReduceMetadata prepare_reduce_metadata;
+
+  ORT_RETURN_IF_ERROR(PrepareForReduce(X,
                                        keepdims_,
                                        axes_,
-                                       &X,
-                                       &Y,
-                                       input_count,
-                                       output_count,
-                                       output_dims,
-                                       input_dims_cudnn,
-                                       output_dims_cudnn,
-                                       rank, stride));
+                                       prepare_reduce_metadata));
+
+  Tensor* Y = ctx->Output(0, prepare_reduce_metadata.squeezed_output_dims);
+
+  int64_t input_count = prepare_reduce_metadata.input_count;
+  int64_t output_count = prepare_reduce_metadata.output_count;
+  std::vector<int64_t>& input_dims_cudnn = prepare_reduce_metadata.input_dims_cudnn;
+  std::vector<int64_t>& output_dims_cudnn = prepare_reduce_metadata.output_dims_cudnn;
+
+  // This reduction keep adding values to this buffer. If a non-zero value, say 1000, is here, the sum will start with 1000.
+  // Therefore zeroing out the memory is required
+  CUDA_RETURN_IF_ERROR(cudaMemset(Y->MutableDataRaw(), 0, Y->SizeInBytes()));
 
```

Comment:
Nit: Could do this after the branch on input_count == output_count where we just do a copy.

Reply:
Not sure what you mean - if input_count == output_count, does it mean we just do a copy ? Should we just defer these kinds of optimizations of other ops to another PR ?

Reply:
A separate PR is fine. If input_count == output_count we're going to overwrite the entire buffer, so there's no point doing a memset upfront is there?

Reply:
You are right. It can be done in a few other places in the implementation. I ll do this in a separate change. I didn't pay too much attention to the implementation (or potential optimizations) in the reduction ops and only focused on refactoring the interfaces so that they can be used from Einsum.